File: backend/src/main/java/com/calendar/model/Reminder.java
package com.calendar.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;

@Entity
@Table(name = "reminders", indexes = {
    @Index(name = "idx_reminder_event", columnList = "event_id"),
    @Index(name = "idx_reminder_time", columnList = "reminder_time"),
    @Index(name = "idx_reminder_status", columnList = "status")
})
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@EntityListeners(AuditingEntityListener.class)
public class Reminder {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "event_id", nullable = false)
    private Event event;

    @Column(nullable = false)
    private LocalDateTime reminderTime;

    @Column(length = 20)
    @Enumerated(EnumType.STRING)
    private ReminderStatus status;

    @Column(length = 500)
    private String message;

    @Column
    private Boolean sent = false;

    @Column
    private LocalDateTime sentAt;

    @Column(length = 36)
    private String notificationId;

    @CreatedDate
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    public enum ReminderStatus {
        PENDING, SENT, CANCELLED, FAILED
    }
}

File: core/domain/src/main/java/com/calendar/core/domain/model/Reminder.kt
package com.calendar.core.domain.model

import java.time.LocalDateTime

/**
 * 提醒数据模型
 */
data class Reminder(
    val id: Long = 0,
    val eventId: Long,
    val reminderTime: LocalDateTime,
    val message: String? = null,
    val status: ReminderStatus = ReminderStatus.PENDING,
    val notificationId: Int? = null,
    val createdAt: LocalDateTime = LocalDateTime.now()
)

enum class ReminderStatus {
    PENDING,    // 待触发
    TRIGGERED,  // 已触发
    CANCELLED,  // 已取消
    FAILED      // 失败
}

/**
 * 提醒规则
 */
data class ReminderRule(
    val minutesBefore: Int,  // 事件开始前多少分钟提醒
    val isEnabled: Boolean = true
) {
    companion object {
        // 预设提醒时间
        val AT_TIME = ReminderRule(0)  // 准时
        val FIVE_MINUTES = ReminderRule(5)
        val FIFTEEN_MINUTES = ReminderRule(15)
        val THIRTY_MINUTES = ReminderRule(30)
        val ONE_HOUR = ReminderRule(60)
        val ONE_DAY = ReminderRule(24 * 60)
    }
}

/**
 * 重复规则
 */
data class RecurrenceRule(
    val type: RecurrenceType,
    val interval: Int = 1,  // 间隔
    val endDate: LocalDateTime? = null,  // 结束日期
    val count: Int? = null  // 重复次数
)

enum class RecurrenceType {
    NONE,       // 不重复
    DAILY,      // 每天
    WEEKLY,     // 每周
    MONTHLY,    // 每月
    YEARLY,     // 每年
    CUSTOM      // 自定义
}
